File: ios/Classes/EaseCallKitPlugin.m
#import "EaseCallKitPlugin.h"
#import <HyphenateChat/HyphenateChat.h>

#define ease_call_dispatch_main_async_safe(block)\
if ([NSThread isMainThread]) {\
block();\
} else {\
dispatch_async(dispatch_get_main_queue(), block);\
}


@interface EaseCallKitPlugin()<EaseCallDelegate>
@property (nonatomic, strong) FlutterMethodChannel* channel;
@end

@implementation EaseCallKitPlugin
+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
    FlutterMethodChannel* channel = [FlutterMethodChannel
                                     methodChannelWithName:@"ease_call_kit"
                                     binaryMessenger:[registrar messenger]];
    EaseCallKitPlugin* instance = [[EaseCallKitPlugin alloc] init];
    instance.channel = channel;
    [registrar addMethodCallDelegate:instance channel:channel];
}


- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
    if ([@"initCallKit" isEqualToString:call.method]) {
        [self initCallKit:call.arguments result:result];
    } else if ([@"startSingleCall" isEqualToString:call.method]) {
        [self startSingleCall:call.arguments result:result];
    } else if ([@"startInviteUsers" isEqualToString:call.method]) {
        [self startInviteUsers:call.arguments result:result];
    } else if ([@"getEaseCallConfig" isEqualToString:call.method]) {
        [self getEaseCallConfig:call.arguments result:result];
    } else if ([@"setRTCToken" isEqualToString:call.method]) {
        [self setRTCToken:call.arguments result:result];
    }else if ([@"setUsersMapper" isEqualToString:call.method]){
        [self setUsersMapper:call.arguments result:result];
    }else if ([@"setUserInfoMapper" isEqualToString:call.method]){
        [self setUsersMapper:call.arguments result:result];
    }else if ([@"getTestUserToken" isEqualToString:call.method]){
        if (EMClient.sharedClient.accessUserToken.length > 0) {
            result(EMClient.sharedClient.accessUserToken);
        }else {
            result(nil);
        }
    }else{
        result(FlutterMethodNotImplemented);
    }
}

- (void)initCallKit:(NSDictionary *)dict result:(FlutterResult)result{
    EaseCallConfig *config = [EaseCallConfig fromJson:dict];
    [[EaseCallManager sharedManager] initWithConfig:config delegate:self];
    result(nil);
}

- (void)startSingleCall:(NSDictionary *)dict result:(FlutterResult)result{
    
    EaseCallType type = [dict[@"call_type"] intValue] == 0 ? EaseCallType1v1Audio : EaseCallType1v1Video;
    NSString *emId = dict[@"em_id"];
    NSDictionary *ext = dict[@"ext"];
    
    [[EaseCallManager sharedManager] startSingleCallWithUId:emId
                                                       type:type
                                                        ext:ext
                                                 completion:^(NSString * _Nonnull callId, EaseCallError * aError)
     {
        if (aError) {
            [self callDidOccurError:aError];
        }
    }];
    ease_call_dispatch_main_async_safe(^(){
        result(nil);
    });
}

- (void)startInviteUsers:(NSDictionary *)dict result:(FlutterResult)result{
    NSArray *users = dict[@"users"];
    NSDictionary *ext = dict[@"ext"];
    [EaseCallManager.sharedManager startInviteUsers:users
                                                ext:ext
                                         completion:^(NSString * _Nullable callId, EaseCallError * _Nullable aError)
     {
        if (aError) {
            [self callDidOccurError:aError];
        }
    }];
    
    ease_call_dispatch_main_async_safe(^(){
        result(nil);
    });
}

- (void)getEaseCallConfig:(NSDictionary *)dict result:(FlutterResult)result{
    EaseCallConfig *config = EaseCallManager.sharedManager.getEaseCallConfig;
    ease_call_dispatch_main_async_safe(^(){
        result([config toJson]);
    });
}

- (void)setRTCToken:(NSDictionary *)dict result:(FlutterResult)result {
    NSString *rtcToken = dict[@"rtc_token"];
    NSString *channelName = dict[@"channel_name"];
    NSInteger uid = [dict[@"uid"] intValue];
    [EaseCallManager.sharedManager setRTCToken:rtcToken channelName:channelName uid:uid];
    ease_call_dispatch_main_async_safe(^(){
        result(nil);
    });
}


- (void)setUsersMapper:(NSDictionary *)dict result:(FlutterResult)result {
    NSDictionary *receiveMap = dict[@"map"];
    NSMutableDictionary *userMap = [NSMutableDictionary dictionary];
    for (NSString *username in receiveMap.allKeys) {
        userMap[receiveMap[username]] = username;
    }
    NSString *channelName = dict[@"channel_name"];
    [EaseCallManager.sharedManager setUsers:userMap channelName:channelName];
    ease_call_dispatch_main_async_safe(^(){
        result(nil);
    });
}

- (void)setUserInfoMapper:(NSDictionary *)dict result:(FlutterResult)result {
    EaseCallConfig *config =  EaseCallManager.sharedManager.getEaseCallConfig;
    if(config != nil) {
        NSMutableDictionary *usersDict = [NSMutableDictionary dictionary];
        NSArray *usersList = dict[@"userInfo_list"];
        for (NSDictionary *userDict in usersList) {
            NSString *key = userDict[@"key"];
            NSDictionary *value = userDict[@"value"];
            usersDict[key] = [EaseCallUser fromJson:value];
        }
        config.users = usersDict;
    }
    ease_call_dispatch_main_async_safe(^(){
        result(nil);
    });
}

#pragma mark - EaseCallDelegate
- (void)callDidEnd:(NSString * _Nonnull)aChannelName
            reason:(EaseCallEndReason)aReason
              time:(int)aTm
              type:(EaseCallType)aType
{
    
    NSMutableDictionary *dict = [NSMutableDictionary dictionary];
    dict[@"channel_name"] = aChannelName;
    dict[@"time"] = @(aTm);
    dict[@"call_type"] = @([self callTypeToInt:aType]);
    dict[@"reason"] = @([self reasonToInt:aReason]);
    [self.channel invokeMethod:@"callDidEnd" arguments:dict];
}

- (void)callDidOccurError:(EaseCallError * _Nonnull)aError
{
    [self.channel invokeMethod:@"callDidOccurError" arguments:[aError toJson]];
}

- (void)callDidReceive:(EaseCallType)aType
               inviter:(NSString * _Nonnull)user
                   ext:(NSDictionary * _Nullable)aExt
{
    NSMutableDictionary *arg = [NSMutableDictionary dictionary];
    arg[@"call_type"] = @([self callTypeToInt:aType]);
    arg[@"inviter"] = user;
    arg[@"ext"] = aExt;
    
    [self.channel invokeMethod:@"callDidReceive" arguments:arg];
}


- (void)multiCallDidInvitingWithCurVC:(UIViewController * _Nonnull)vc
                         excludeUsers:(NSArray<NSString *> * _Nullable)users
                                  ext:(NSDictionary * _Nullable)aExt
{
    NSMutableDictionary *arg = [NSMutableDictionary dictionary];
    arg[@"exclude_users"] = users;
    arg[@"ext"] = aExt;
    [self.channel invokeMethod:@"multiCallDidInviting" arguments:arg];
}

- (void)callDidRequestRTCTokenForAppId:(NSString * _Nonnull)aAppId
                           channelName:(NSString * _Nonnull)aChannelName
                               account:(NSString * _Nonnull)aUserAccount
                                   uid:(NSInteger)aAgoraUid {
    NSMutableDictionary *arg = [NSMutableDictionary dictionary];
    arg[@"app_id"] = aAppId;
    arg[@"channel_name"] = aChannelName;
    arg[@"account"] = aUserAccount;
      
    [self.channel invokeMethod:@"callDidRequestRTCToken" arguments:arg];
}

- (void)remoteUserDidJoinChannel:(NSString *)aChannelName uid:(NSInteger)aAgoraUid username:(NSString *)aUserName {
    NSMutableDictionary *arg = [NSMutableDictionary dictionary];
    arg[@"channel_name"] = aChannelName;
    arg[@"account"] = aUserName;
    arg[@"agora_uid"] = @(aAgoraUid);
    
    [self.channel invokeMethod:@"remoteUserDidJoinChannel" arguments:arg];
}

- (void)callDidJoinChannel:(NSString * _Nonnull)aChannelName uid:(NSUInteger)aAgoraUid {
    NSMutableDictionary *arg = [NSMutableDictionary dictionary];
    arg[@"channel_name"] = aChannelName;
    arg[@"agora_uid"] = @(aAgoraUid);
    [self.channel invokeMethod:@"callDidJoinChannel" arguments:arg];
}


#pragma mark private method

- (int)callTypeToInt:(EaseCallType)aType {
    int type = 1;
    switch (aType) {
        case EaseCallType1v1Audio:
            type = 1;
            break;
        case EaseCallType1v1Video:
            type = 1;
            break;
        case EaseCallTypeMulti:
            type = 1;
            break;
        default:
            break;
    }
    return type;
}

- (int)reasonToInt:(EaseCallEndReason)aReason{
    int reason = 1;
    switch (aReason) {
        case EaseCallEndReasonHangup:
            reason = 1;
            break;
            
        case EaseCallEndReasonCancel:
            reason = 2;
            break;
            
        case EaseCallEndReasonRemoteCancel:
            reason = 3;
            break;
            
        case EaseCallEndReasonRefuse:
            reason = 4;
            break;
            
        case EaseCallEndReasonBusy:
            reason = 5;
            break;
            
        case EaseCallEndReasonNoResponse:
            reason = 6;
            break;
            
        case EaseCallEndReasonRemoteNoResponse:
            reason = 7;
            break;
            
        case EaseCallEndReasonHandleOnOtherDevice:
            reason = 8;
            break;
            
        default:
            break;
    }
    
    return reason;
}


@end

@implementation EaseCallError (Flutter)

- (NSDictionary *)toJson {
    NSMutableDictionary *ret = [NSMutableDictionary dictionary];
    ret[@"err_code"] = @(self.errCode);
    ret[@"err_desc"] = self.errDescription;
    ret[@"err_type"] = ({
        int type = 0;
        if (self.aErrorType == EaseCallErrorTypeProcess) {
            type = 1;
        }else if(self.aErrorType == EaseCallErrorTypeRTC) {
            type = 2;
        }else if(self.aErrorType == EaseCallErrorTypeIM) {
            type = 3;
        }
        @(type);
    });
    return ret;
}
@end


@implementation EaseCallConfig (Flutter)

+ (EaseCallConfig *)fromJson:(NSDictionary *)dict {
    EaseCallConfig *config = [[EaseCallConfig alloc] init];
    config.agoraAppId = dict[@"agora_app_id"];
    config.defaultHeadImage = dict[@"default_head_image_url"];
    if ([dict[@"call_timeout"] intValue] > 0) {
        config.callTimeOut = [dict[@"call_timeout"] intValue];
    }
    
    NSString *ringFileURL = dict[@"ring_file_url"];
    if (ringFileURL.length > 0) {
        config.ringFileUrl = [NSURL URLWithString:ringFileURL];
    }
    
    config.enableRTCTokenValidate = [dict[@"enable_rtc_token_validate"] boolValue];
    
    NSMutableDictionary *usersDict = [NSMutableDictionary dictionary];
    NSArray *usersList = dict[@"user_map"];
    for (NSDictionary *userDict in usersList) {
        NSString *key = userDict[@"key"];
        NSDictionary *value = userDict[@"value"];
        usersDict[key] = [EaseCallUser fromJson:value];
    }
    config.users = usersDict;
    return config;
}

- (NSDictionary *)toJson {
    NSMutableDictionary *ret = [NSMutableDictionary dictionary];
    ret[@"agora_app_id"] = self.agoraAppId;
    ret[@"default_head_image_url"] = self.defaultHeadImage;
    ret[@"call_timeout"] = @(self.callTimeOut);
    ret[@"ring_file_url"] = self.ringFileUrl;
    ret[@"enable_rtc_token_validate"] = @(self.enableRTCTokenValidate);
    NSMutableDictionary *usersDict = [NSMutableDictionary dictionary];
    for (NSString *key in self.users.allKeys) {
        EaseCallUser *user = usersDict[key];
        usersDict[key] = [user toJson];
    }
    ret[@"user_map"] = usersDict;
    return ret;
}
@end

@implementation EaseCallUser (Flutter)

+ (EaseCallUser *)fromJson:(NSDictionary *)dict {
    EaseCallUser *user = [[EaseCallUser alloc] init];
    NSString *avatarStr = dict[@"avatar_url"];
    if (avatarStr) {
        user.headImage = [NSURL URLWithString:avatarStr];
    }
    NSString *nickname = dict[@"nickname"];
    if (nickname) {
        user.nickName = nickname;
    }
    
    return user;
}

- (NSDictionary *)toJson {
    NSMutableDictionary *ret = [NSMutableDictionary dictionary];
    ret[@"avatar_url"] = self.headImage.absoluteString;
    ret[@"nickname"] = self.nickName;
    return ret;
}

@end

File: android/src/main/java/com/easemob/callkit/ease_call_kit/EaseCallKitPlugin.java
package com.easemob.callkit.ease_call_kit;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.Intent;
import android.os.AsyncTask;
import android.os.Handler;
import android.os.Looper;
import android.os.Message;
import android.util.Pair;

import com.hyphenate.EMValueCallBack;
import com.hyphenate.chat.EMClient;
import com.hyphenate.chat.EMOptions;
import com.hyphenate.chat.EMUserInfo;
import com.hyphenate.cloud.EMHttpClient;
import com.hyphenate.easecallkit.EaseCallKit;
import com.hyphenate.easecallkit.base.EaseCallEndReason;
import com.hyphenate.easecallkit.base.EaseCallKitConfig;
import com.hyphenate.easecallkit.base.EaseCallKitListener;
import com.hyphenate.easecallkit.base.EaseCallKitTokenCallback;
import com.hyphenate.easecallkit.base.EaseCallType;
import com.hyphenate.easecallkit.base.EaseCallUserInfo;
import com.hyphenate.easecallkit.base.EaseGetUserAccountCallback;
import com.hyphenate.easecallkit.base.EaseUserAccount;
import com.hyphenate.exceptions.HyphenateException;
import com.hyphenate.util.EMLog;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map;

import androidx.annotation.NonNull;

import io.flutter.Log;
import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.plugin.common.JSONUtil;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;



import static android.content.Intent.FLAG_ACTIVITY_NEW_TASK;

/** EaseCallKitPlugin */
public class EaseCallKitPlugin implements FlutterPlugin, MethodCallHandler {

  private MethodChannel channel;
  private FlutterPluginBinding binding;
  private WeakReference<EaseCallKitTokenCallback> weakCallback;
  private Map<String, EaseCallKitTokenCallback> tokenCallbackMap;
  private Map<String, EaseGetUserAccountCallback> userAccountCallbackMap;

  private EaseCallKitListener callKitListener;

  static final Handler handler = new Handler(Looper.getMainLooper());

  public void post(Runnable runnable) {
    handler.post(runnable);
  }

  @Override
  public void onAttachedToEngine(@NonNull FlutterPluginBinding flutterPluginBinding) {
    binding = flutterPluginBinding;
    channel = new MethodChannel(flutterPluginBinding.getBinaryMessenger(), "ease_call_kit");
    channel.setMethodCallHandler(this);
    addCallKitListener();
  }

  @Override
  public void onMethodCall(@NonNull MethodCall call, @NonNull Result result) {

    JSONObject param = null;
    if (call.arguments != null) {
      param = new JSONObject((Map) call.arguments);
    }
    try {
      if (call.method.equals("initCallKit")) {
        initWithConfig(param, result);
      } else if (call.method.equals("startSingleCall")) {
        startSingleCall(param, result);
      } else if (call.method.equals("startInviteUsers")) {
        startInviteUsers(param, result);
      } else if (call.method.equals("getEaseCallConfig")) {
        getEaseCallConfig(param, result);
      } else if (call.method.equals("setRTCToken")) {
        setRTCToken(param, result);
      } else if (call.method.equals("setUsersMapper")){
        setUsersMapper(param, result);
      } else if (call.method.equals("setUserInfoMapper")){
        setUserInfoMapper(param, result);
      } else if (call.method.equals("getTestUserToken")){
        if (EMClient.getInstance().getAccessToken().length() > 0) {
          result.success(EMClient.getInstance().getAccessToken());
        }else {
          result.success(null);
        }
      } else {
        result.notImplemented();
      }
    } catch (JSONException e) {
      result.notImplemented();
    }
  }

  private void initWithConfig(JSONObject map, final Result result) throws JSONException {
    EaseCallKit.getInstance().init(this.binding.getApplicationContext(), configFromJson(map));
    post(new Runnable() {
      @Override
      public void run() {
        result.success(null);
      }
    });
  }


  private void startSingleCall(JSONObject map, final Result result) throws JSONException {
    EaseCallType callType = map.getInt("call_type") == 0 ? EaseCallType.SINGLE_VOICE_CALL : EaseCallType.SINGLE_VIDEO_CALL;
    String user = map.getString("em_id");
    Map<String,Object> ext = null;
    if (map.has("ext")){
      ext = JsonObjectToHashMap(map.getJSONObject("ext"));
    }
    EaseCallKit.getInstance().startSingleCall(callType, user, ext);
    post(new Runnable() {
      @Override
      public void run() {
        result.success(null);
      }
    });
  }

  private void startInviteUsers(JSONObject map, final Result result) throws JSONException {
    JSONArray usersArray = map.getJSONArray("users");
    String[] users = new String[usersArray.length()];
    for (int i = 0; i < usersArray.length(); i++) {
      users[i] = (String) usersArray.get(i);
    }
    Map<String,Object> ext = null;
    if (map.has("ext")){
      ext = JsonObjectToHashMap(map.getJSONObject("ext"));
    }
    EaseCallKit.getInstance().startInviteMultipleCall(users, ext);
    post(new Runnable() {
      @Override
      public void run() {
        result.success(null);
      }
    });
  }

  // TODO: 不能reset，获取感觉也没啥用是否需要提供?
  private void getEaseCallConfig(JSONObject map, Result result) {
      EaseCallKitConfig config = EaseCallKit.getInstance().getCallKitConfig();
  }

  private void setRTCToken(JSONObject map, final Result result) throws JSONException {
    String rtcToken = map.getString("rtc_token");
    String channelName = map.getString("channel_name");
    Integer uid = map.getInt("uid");
    if (tokenCallbackMap.containsKey(channelName)) {
       EaseCallKitTokenCallback callback = tokenCallbackMap.get(channelName);
       callback.onSetToken(rtcToken, uid);
       tokenCallbackMap.remove(callback);
    }else{
        // ??
    }
    // 是否需要根据onSetToken/onGetTokenError分别返回？
    post(new Runnable() {
      @Override
      public void run() {
        result.success(null);
      }
    });
  }


  private void setUsersMapper(JSONObject map, final Result result) throws JSONException {
    JSONObject userMap = map.getJSONObject("map");
    String channelName = map.getString("channel_name");
    if(userAccountCallbackMap.containsKey(channelName)) {
        EaseGetUserAccountCallback callback = userAccountCallbackMap.get(channelName);
        List<EaseUserAccount> list = new ArrayList<>();
      Iterator<String> it = userMap.keys();
      while(it.hasNext()){
        String key = it.next();
        int value = userMap.getInt(key);
        EaseUserAccount account = new EaseUserAccount(value, key);
        list.add(account);
      }
      callback.onUserAccount(list);
      userAccountCallbackMap.remove(callback);
    }else {
      // ??
    }
    post(new Runnable() {
      @Override
      public void run() {
        result.success(null);
      }
    });
  }

  private void setUserInfoMapper(JSONObject map, final Result result) throws JSONException {
    JSONArray userInfoList = map.getJSONArray("userInfo_list");
    Map<String, EaseCallUserInfo> userMap = new HashMap<>();
    for (int i = 0; i < userInfoList.length(); i++) {
      JSONObject userJsonObj = userInfoList.getJSONObject(i);
      String username = userJsonObj.getString("key");
      JSONObject accountJsonObj = userJsonObj.getJSONObject("value");
      String avatarUrl = null;
      if (accountJsonObj.has("avatar_url")) {
        avatarUrl = accountJsonObj.getString("avatar_url");
      }
      String nickname = null;
      if (accountJsonObj.has("nickname")) {
        nickname = accountJsonObj.getString("nickname");
      }
      userMap.put(username, new EaseCallUserInfo(nickname, avatarUrl));
    }

    EaseCallKitConfig config = EaseCallKit.getInstance().getCallKitConfig();
    config.setUserInfoMap(userMap);
    post(new Runnable() {
      @Override
      public void run() {
        result.success(null);
      }
    });
  }

  private void addCallKitListener() {
    callKitListener = new EaseCallKitListener() {
      @Override
      public void onInviteUsers(Context context, String userId[], JSONObject ext) {
        Map<String, Object> data = new HashMap<>();
        List<String> users = new ArrayList<>();
        for (int i = 0; i < userId.length; i++) {
          users.add(userId[i]);
        }
        data.put("exclude_users", users);

        Map<String, Object> extMap = null;
          try{
            if (ext != null) {
              extMap = JsonObjectToHashMap(ext);
            }
          }catch(JSONException e) {
          }finally {
            if (extMap != null) {
              data.put("ext", extMap);
            }
            channel.invokeMethod("multiCallDidInviting", data);
          }
      }

      @Override
      public void onEndCallWithReason(EaseCallType callType, String channelName, EaseCallEndReason reason,
          long callTime) {
        final Map<String, Object> data = new HashMap<>();
        data.put("channel_name", channelName);
        data.put("time", callTime);
        data.put("call_type", callTypeToInt(callType));
        data.put("reason", reasonToInt(reason));
        post(new Runnable() {
          @Override
          public void run() {
            channel.invokeMethod("callDidEnd", data);
          }
        });
      }

      @Override
      public void onGenerateToken(String userId, String channelName, String agoraAppId, EaseCallKitTokenCallback callback) {

        if (tokenCallbackMap == null) {
          tokenCallbackMap = new HashMap<>();
        }
        tokenCallbackMap.put(channelName, callback);
        final Map<String, Object> data = new HashMap<>();
        data.put("app_id", agoraAppId);
        data.put("channel_name", channelName);
        data.put("account", userId);

        post(new Runnable() {
          @Override
          public void run() {
            channel.invokeMethod("callDidRequestRTCToken", data);
          }
        });
      }


      @Override
      public void onReceivedCall(EaseCallType callType, String fromUserId, JSONObject ext)  {
        final Map<String, Object> data = new HashMap<>();
        data.put("call_type", callTypeToInt(callType));
        data.put("inviter", fromUserId);

        if (ext != null && ext.length() > 0) {
          String  jsonString = ext.toString();
          data.put("ext", jsonString);
        }

        post(new Runnable() {
          @Override
          public void run() {
            channel.invokeMethod("callDidReceive", data);
          }
        });
      }

      @Override
      public void onCallError(EaseCallKit.EaseCallError type, int errorCode, String description) {

        final Map<String, Object> data = new HashMap<>();
        int intType = 0;
        if (type == EaseCallKit.EaseCallError.PROCESS_ERROR) {
          intType = 1;
        } else if (type == EaseCallKit.EaseCallError.RTC_ERROR) {
          intType = 2;
        } else {
          intType = 3;
        }
        data.put("err_type", intType);
        data.put("err_code", errorCode);
        data.put("err_desc", description);

        post(new Runnable() {
          @Override
          public void run() {
            channel.invokeMethod("callDidOccurError", data);
          }
        });
      }

      @Override
      public void onInViteCallMessageSent() {

      }

      @Override
      public void onRemoteUserJoinChannel(String channelName, String userName, int uid, EaseGetUserAccountCallback callback) {
        if (userAccountCallbackMap == null) {
          userAccountCallbackMap = new HashMap<>();
        }

        userAccountCallbackMap.put(channelName, callback);

        final Map<String, Object> data = new HashMap<>();
        data.put("channel_name", channelName);
        data.put("account", userName);
        data.put("agora_uid", uid);

        post(new Runnable() {
          @Override
          public void run() {
            channel.invokeMethod("remoteUserDidJoinChannel", data);
          }
        });
      }
    };

    EaseCallKit.getInstance().setCallKitListener(callKitListener);
  }



  private static HashMap<String, Object> JsonObjectToHashMap(JSONObject data) throws JSONException {
    HashMap<String, Object> map = new HashMap<String,Object>();
    Iterator iterator = data.keys();
    while (iterator.hasNext()) {
      String key = iterator.next().toString();
      Object result = data.get(key);
      if (result.getClass().getSimpleName().equals("Integer")) {
        map.put(key, (Integer) result);
      } else if (result.getClass().getSimpleName().equals("Boolean")) {
        map.put(key, (Boolean) result);
      } else if (result.getClass().getSimpleName().equals("Long")) {
        map.put(key, (Long) result);
      } else if (result.getClass().getSimpleName().equals("JSONObject")) {
        map.put(key, (JSONObject) result);
      } else if (result.getClass().getSimpleName().equals("JSONArray")) {
        map.put(key, (JSONArray) result);
      } else {
        map.put(key, data.getString(key));
      }
    }
    return map;
  }

  private EaseCallKitConfig configFromJson(JSONObject map) throws JSONException {
    EaseCallKitConfig config = new EaseCallKitConfig();
    config.setAgoraAppId(map.getString("agora_app_id"));

    if (map.has("default_head_image_url")) {
      config.setDefaultHeadImage(map.getString("default_head_image_url"));
    }

    if (map.has("call_timeout")) {
      config.setCallTimeOut(map.getInt("call_timeout") * 1000);
    }

    if (map.has("ring_file_url")) {
      config.setRingFile(map.getString("ring_file_url"));
    }

    config.setEnableRTCToken(map.has("enable_rtc_token_validate") ? map.getBoolean("enable_rtc_token_validate") : true);
    Map<String, EaseCallUserInfo> userMap = new HashMap<>();
    if (map.has("user_map")) {
      JSONArray usersList = map.getJSONArray("user_map");
      for (int i = 0; i < usersList.length(); i++) {
        JSONObject userObject = (JSONObject) usersList.get(i);
        EaseCallUserInfo userInfo = userInfoFromJson(userObject.getJSONObject("value"));
        userMap.put(userObject.getString("key"), userInfo);
      }
      config.setUserInfoMap(userMap);
    }

    return config;
  }

  private EaseCallUserInfo userInfoFromJson(JSONObject map) throws JSONException {

    String name = null;
    if(map.has("nickname")) {
      name = map.getString("nickname");
    }

    String headImage = null;
    if (map.has("avatar_url")) {
      headImage = map.getString("avatar_url");
    }

    EaseCallUserInfo userInfo = new EaseCallUserInfo(name, headImage);

    return userInfo;
  }

  private int callTypeToInt(EaseCallType type) {
    int intType = 1;
    if (type == EaseCallType.SINGLE_VOICE_CALL) {
      intType = 1;
    } else if (type == EaseCallType.SINGLE_VIDEO_CALL) {
      intType = 2;
    } else {
      intType = 3;
    }
    return intType;
  }

  private int reasonToInt(EaseCallEndReason reason) {
    int intReason = 1;
    switch (reason) {
    case EaseCallEndReasonHangup:
      intReason = 1;
      break;
    case EaseCallEndReasonCancel:
      intReason = 2;
      break;
    case EaseCallEndReasonRemoteCancel:
      intReason = 3;
      break;
    case EaseCallEndReasonRefuse:
      intReason = 4;
      break;
    case EaseCallEndReasonBusy:
      intReason = 5;
      break;
    case EaseCallEndReasonNoResponse:
      intReason = 6;
      break;
    case EaseCallEndReasonRemoteNoResponse:
      intReason = 7;
      break;
    case EaseCallEndReasonHandleOnOtherDevice:
      intReason = 8;
      break;
    default:
      break;
    }
    return intReason;
  }

  @Override
  public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
    channel.setMethodCallHandler(null);
  }
}
